File: Day49/Part2/arcane_fusion.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

long long solve_arcane_fusion(const vector<long long>& nums) {
    vector<long long> basis(64, 0);
    
    // Build Linear Basis
    for (long long num : nums) {
        for (int i = 63; i >= 0; --i) {
            if ((num >> i) & 1) {
                if (basis[i] == 0) {
                    basis[i] = num;
                    break;
                } else {
                    num ^= basis[i];
                }
            }
        }
    }
    
    // Construct Maximum XOR
    long long max_xor = 0;
    for (int i = 63; i >= 0; --i) {
        if ((max_xor ^ basis[i]) > max_xor) {
            max_xor ^= basis[i];
        }
    }
    
    return max_xor;
}

int main() {
    vector<long long> nums = {3, 10, 5, 25, 2, 8};
    cout << "Maximum XOR: " << solve_arcane_fusion(nums) << endl;
    return 0;
}

File: learnLeetCode/Day49/Part2/arcane_fusion.py
def solve_arcane_fusion(nums):
    basis = [0] * 64
    
    # Build the Linear Basis
    for num in nums:
        for i in range(63, -1, -1):
            if (num >> i) & 1:
                if basis[i] == 0:
                    basis[i] = num
                    break
                else:
                    num ^= basis[i]
                    
    # Construct Maximum XOR
    max_xor = 0
    for i in range(63, -1, -1):
        if (max_xor ^ basis[i]) > max_xor:
            max_xor ^= basis[i]
            
    return max_xor

if __name__ == "__main__":
    nums = [3, 10, 5, 25, 2, 8]
    result = solve_arcane_fusion(nums)
    print(f"Maximum XOR: {result}")
